EventListener/LeadTimelineSubscriber.php

```diff
@@ -0,0 +1,142 @@
+        // but also should query for a total number of events and return an array of ['total' => $x, 'results' => []].
+        // There is a TimelineTrait to assist with this. See repository example.$repo         = $this->em->getRepository('MauticContactSourceBundle:Event');
+        $stats = $repo->getTimelineStats($event->getLeadId(), $event->getQueryOptions());
+
+        // If isEngagementCount(), this event should only inject $stats into addToCounter() to append to data to generate
+        // the engagements graph. Not all events are engagements if they are just informational so it could be that this
+        // line should only be used when `!$event->isEngagementCount()`. Using TimelineTrait will determine the appropriate
+        // return value based on the data included in getQueryOptions() if used in the stats method above.
+        // $event->addToCounter($eventTypeKey, $stats);
+
+        if (!$event->isEngagementCount()) {
+            // Add the events to the event array
+            foreach ($stats['results'] as $stat) {
+                if ($stat['timestamp']) {
+                    $event->addEvent(
+                        [
+                            // Event key type
+                            'event'           => $eventTypeKey,
+                            // Event name/label - can be a string or an array as below to convert to a link
+                            'eventLabel'      => [
+                                'label'  => ucwords(
+                                        $stat['enhancer']
+                                    ).(isset($stat['location']) ? ' / '.$stat['location'] : ''),
+                                'href'   => (!empty($stat['share_url']) ? $stat['share_url'] : '#'),
+                                'target' => '_blank',
+                            ],
+                            // Translated string displayed in the Event Type column
+                            'eventType'       => $this->translator->trans(
+                                'mautic.enhancer.integration.'.strtolower($stat['enhancer']).'.status.'.$stat['status']
+                            ),
+                            'timestamp'       => $stat['timestamp'],
+                            'details'         => $stat,
+                            'extra'           => [
+                                'raw' => self::describeStat($stat),
+                            ],
+                            'contentTemplate' => 'MauticEnhancerBundle:Timeline:enhancer.html.php',
+                            'icon'            => 'fa-plus-square-o enhancer-button',
+                        ]
+                    );
+                }
+            }
+        }
+    }
+
+    /**
+     * Moved here to pass tests and still support recursive enhancer data.
+     *
+     * @param     $value
+     * @param int $depth
+     *
+     * @return string
+     */
+    private static function describeStat($value, $depth = 1)
```

Comment:
Is there any reason to make it static?

Reply:
Just a pattern I've been using a lot lately. If a helper method is isolated as to never need $this, it should generally be a static method to enforce that continued isolated scope. It keeps things from becoming spaghetti as a class grows and makes it easier to divide later if needed.